fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/PredicateCountByLevelProperty.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * PredicateCountByLevelProperty.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.properties;
+
+import com.apple.foundationdb.record.query.plan.cascades.ExpressionProperty;
+import com.apple.foundationdb.record.query.plan.cascades.Reference;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpressionVisitor;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpressionVisitorWithDefaults;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpressionWithPredicates;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class PredicateCountByLevelProperty implements ExpressionProperty<PredicateCountByLevelProperty.PredicateCountByLevelInfo> {
+    @Nonnull
+    private static final PredicateCountByLevelProperty PREDICATE_COUNT_BY_LEVEL = new PredicateCountByLevelProperty();
+
+    private PredicateCountByLevelProperty() {
+        // prevent outside instantiation
+    }
+
+    @Nonnull
+    @Override
+    public RelationalExpressionVisitor<PredicateCountByLevelInfo> createVisitor() {
+        return PredicateCountByLevelVisitor.VISITOR;
+    }
+
+    public PredicateCountByLevelInfo evaluate(RelationalExpression expression) {
+        return Objects.requireNonNull(createVisitor().visit(expression));
+    }
+
+    @Nonnull
+    public static PredicateCountByLevelProperty predicateCountByLevel() {
+        return PREDICATE_COUNT_BY_LEVEL;
+    }
+
+    public static final class PredicateCountByLevelInfo {
+        private final Map<Integer, Integer> levelToPredicateCount;
+        private final int highestLevel;
+
+        public PredicateCountByLevelInfo(Map<Integer, Integer> levelToPredicateCount, int highestLevel) {
```

Comment:
Everything `public` needs a javadoc. Will not annotate stuff like that everywhere. Maybe you can go through it and just add that when you one missing.